# Цели исследования:
Проанализировать базу данных дял того что бы лучше формулировать предложения для покупателей

### Задания

- Посчитайте, сколько книг вышло после 1 января 2000 года;
- Для каждой книги посчитайте количество обзоров и среднюю оценку;
- Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;
- Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;
- Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

### Описание данных

**Таблица `books`**

Содержит данные о книгах:

- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

**Таблица `authors`**

Содержит данные об авторах:

- `author_id` — идентификатор автора;
- `author` — имя автора.

**Таблица `publishers`**

Содержит данные об издательствах:

- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

**Таблица `ratings`**

Содержит данные о пользовательских оценках книг:

- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

**Таблица `reviews`**

Содержит данные о пользовательских обзорах на книги:

- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.

## Подключение:

In [1]:
#импортируем библиотеки
import pandas as pd 
from sqlalchemy import create_engine 
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
                'pwd': 'Sdf4$2;d-d30pp', # пароль
                'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net', 
                'port': 6432, # порт подключения
                'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'], 
                                                         db_config['pwd'], 
                                                         db_config['host'], 
                                                         db_config['port'], 
                                                         db_config['db']) 
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'}) 

## Изучение таблиц:

In [2]:
query = '''
SELECT 
      *
FROM books
LIMIT 10
'''
pd.io.sql.read_sql(query, con = engine)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   
5        6        257               1st to Die (Women's Murder Club  #1)   
6        7        258               2nd Chance (Women's Murder Club  #2)   
7        8        260              4th of July (Women's Murder Club  #4)   
8        9        563                                   A Beautiful Mind   
9       10        445                                 A Bend in the Road   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268  
5        424       2005-05-20           116  
6        400       2005-05-20           116  
7        448       2006-06-01           318  
8        461       2002-02-04           104  
9        341       2005-04-01           116

In [3]:
query = '''
SELECT 
      *
FROM authors
LIMIT 10
'''
pd.io.sql.read_sql(query, con = engine)

author_id                                author
0          1                            A.S. Byatt
1          2        Aesop/Laura Harris/Laura Gibbs
2          3                       Agatha Christie
3          4                         Alan Brennert
4          5              Alan Moore/David   Lloyd
5          6                            Alan Paton
6          7           Albert Camus/Justin O'Brien
7          8                         Aldous Huxley
8          9    Aldous Huxley/Christopher Hitchens
9         10  Aleksandr Solzhenitsyn/H.T. Willetts

In [4]:
query = '''
SELECT 
      *
FROM publishers
LIMIT 10
'''
pd.io.sql.read_sql(query, con = engine)

publisher_id                                publisher
0             1                                      Ace
1             2                                 Ace Book
2             3                                Ace Books
3             4                            Ace Hardcover
4             5        Addison Wesley Publishing Company
5             6                                  Aladdin
6             7                       Aladdin Paperbacks
7             8                             Albin Michel
8             9                          Alfred A. Knopf
9            10  Alfred A. Knopf Books for Young Readers

In [5]:
query = '''
SELECT 
      *
FROM ratings
LIMIT 10
'''
pd.io.sql.read_sql(query, con = engine)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2
5          6        3  johnsonamanda       4
6          7        3    scotttamara       5
7          8        3    lesliegibbs       5
8          9        4    abbottjames       5
9         10        4   valenciaanne       4

In [6]:
query = '''
SELECT 
      *
FROM reviews
LIMIT 10
'''
pd.io.sql.read_sql(query, con = engine)

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   
5          6        3    lesliegibbs   
6          7        4   valenciaanne   
7          8        4    abbottjames   
8          9        5        npowers   
9         10        5        staylor   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...  
5           Analysis no several cause international.  
6  One there cost another. Say type save. With pe...  
7  Within enough mother. There at system full rec...  
8  Thank now focus realize economy focus fly. Ite...  
9  Game push lot reduce where remember. Including...

## Выполнение заданий:

**1 Посчитайте, сколько книг вышло после 1 января 2000 года.**

In [7]:
table_1 = ''' 
SELECT COUNT(*) 
FROM books
WHERE publication_date > '2000-1-1'; 
''' 

request1 = pd.io.sql.read_sql(table_1, con = engine)
request1

count
0    819

После 1 января 2000 года вышло 819 книг

**2 Для каждой книги посчитайте количество обзоров и среднюю оценку.**

In [18]:
table_2 = '''
SELECT 
    books.title AS title,
    books.book_id AS id,
    COUNT(DISTINCT review_id) AS cnt_rev,
    AVG(rating) AS rating_avg
    
FROM
    books
LEFT JOIN reviews ON books.book_id = reviews.book_id   
LEFT JOIN ratings ON books.book_id = ratings.book_id

GROUP BY
     
     books.book_id
order by
     rating_avg DESC,
     cnt_rev DESC
'''

request2 = pd.io.sql.read_sql(table_2, con = engine)
request2

book_id                                              title  cnt      mean
0        948                            Twilight (Twilight  #1)    7  3.662500
1        302  Harry Potter and the Prisoner of Azkaban (Harr...    6  4.414634
2        299  Harry Potter and the Chamber of Secrets (Harry...    6  4.287500
3        656                                     The Book Thief    6  4.264151
4        734                                   The Glass Castle    6  4.206897
..       ...                                                ...  ...       ...
995      191  Disney's Beauty and the Beast (A Little Golden...    0  4.000000
996      387                               Leonardo's Notebooks    0  4.000000
997      221                          Essential Tales and Poems    0  4.000000
998       83    Anne Rice's The Vampire Lestat: A Graphic Novel    0  3.666667
999      808                            The Natural Way to Draw    0  3.000000

[1000 rows x 4 columns]

Посчитано.  Книга Twilight (Twilight #1) имеет больше всего обзоров - 7.

**3 Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры.**

In [9]:
table_3 = ''' 
SELECT
    A.publisher_id,
    A.publisher,
    A.books
FROM
    (SELECT 
        books.publisher_id AS publisher_id,
        COUNT(books.book_id) AS books,
        publishers.publisher AS publisher 
    FROM 
        books 
        LEFT JOIN publishers ON books.publisher_id=publishers.publisher_id

    WHERE 
        num_pages>50
    GROUP BY
        books.publisher_id,
        publishers.publisher
    ORDER BY 
        books DESC) AS A
LIMIT 1; 
''' 

request3 = pd.io.sql.read_sql(table_3, con = engine)
request3

publisher_id      publisher  books
0           212  Penguin Books     42

Издательство выпустившее больше всего книг с толщиной более 50 страниц Penguin Books.

**4 Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками.**

In [29]:
table_4 = ''' 

WITH

    t as

    (SELECT title,
            COUNT(rating) as cnt_r,
            AVG(rating) as avg_r
     FROM books as bu
     LEFT JOIN ratings as rat ON bu.book_id = rat.book_id
     GROUP BY title)

     SELECT author,
            AVG(avg_r) avg_rating
    FROM t as t
    LEFT JOIN books ON t.title = books.title
    LEFT JOIN author ON books.author_id = author.author_id

    WHERE cnt_r  >= 50
    GROUP BY author
    ORDER BY avg_rating DESC;
''' 

request4 = pd.io.sql.read_sql(table_4, con = engine)
request4

author  avg_rating
0                          J.K. Rowling/Mary GrandPré    4.283844
1                   Markus Zusak/Cao Xuân Việt Khương    4.264151
2                                      J.R.R. Tolkien    4.258446
3                                   Louisa May Alcott    4.192308
4                                       Arthur Golden    4.107143
5                                        Rick Riordan    4.080645
6                                     William Golding    3.901408
7                                       J.D. Salinger    3.825581
8            Paulo Coelho/Alan R. Clarke/Özdemir İnce    3.789474
9   William Shakespeare/Paul Werstine/Barbara A. M...    3.787879
10                                          Dan Brown    3.754540
11                                         Lois Lowry    3.750000
12            George Orwell/Boris Grabnar/Peter Škerl    3.729730
13                                    Stephenie Meyer    3.662500
14                                     John Steinbeck    3.622951

Самая высокая средняя оценка книг у автора J.K. Rowling/Mary GrandPré. на втором месте Markus Zusak/Cao Xuân Việt Khương.

**5 Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.**

In [11]:
table_5 = ''' 

SELECT 
    ROUND(avg(count)) 
FROM
(SELECT
    COUNT(text)
FROM 
    reviews
FULL OUTER JOIN (SELECT username,
                COUNT(rating_id) as cnt
FROM 
    ratings
GROUP BY
        username) AS rating_cnt
ON rating_cnt.username = reviews.username
where cnt > 50
GROUP BY reviews.username) as count;                                  
''' 

request5 = pd.io.sql.read_sql(table_5, con = engine)
request5

round
0   24.0

Среднее количество обзоров от пользвоателей, которые поставили более 50 оценок составляет 24 обзора.

## Выводы:

- После 1 января 2000 года вышло 819 книг
- Книга Twilight (Twilight #1) имеет больше всего обзоров - 7.
- Издательство выпустившее больше всего книг с толщиной более 50 страниц Penguin Books.
- Самая высокая средняя оценка книг у автора J.K. Rowling/Mary GrandPré. на втором месте Markus Zusak/Cao Xuân Việt Khương.
- Среднее количество обзоров от пользвоателей, которые поставили более 50 оценок составляет 24 обзора.